<a href="https://colab.research.google.com/github/LauraJDiaz/Ciencia-De-Datos/blob/main/Prte12_mayo_LosSantos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')
df = pd.read_excel("/content/drive/MyDrive/censo.xlsx")
df


Mounted at /content/drive


cod_dpto Departamento  cod_mpio   Municipio              fec_paquete  \
0            68    SANTANDER     68418  LOS SANTOS  2023-05-03 17:09:24.893   
1            68    SANTANDER     68418  LOS SANTOS  2023-05-03 17:09:24.893   
2            68    SANTANDER     68418  LOS SANTOS  2023-05-03 17:09:24.893   
3            68    SANTANDER     68418  LOS SANTOS  2023-05-03 17:09:24.893   
4            68    SANTANDER     68418  LOS SANTOS  2023-05-03 17:09:24.893   
...         ...          ...       ...         ...                      ...   
12982        68    SANTANDER     68418  LOS SANTOS  2023-12-06 17:24:40.143   
12983        68    SANTANDER     68418  LOS SANTOS  2023-12-06 17:24:40.143   
12984        68    SANTANDER     68418  LOS SANTOS  2023-12-06 17:24:40.143   
12985        68    SANTANDER     68418  LOS SANTOS  2023-12-06 17:24:40.143   
12986        68    SANTANDER     68418  LOS SANTOS  2023-12-06 17:24:40.143   

       num_paquete  num_ficha  ide_ficha_origen  ide_edificacion  \
0             1687     128326      6.841800e+19                0   
1             1687     128326      6.841800e+19                0   
2             1687     128326      6.841800e+19                0   
3             1687     128326      6.841800e+19                0   
4             1687     128326      6.841800e+19                0   
...            ...        ...               ...              ...   
12982         1833     143332      6.841801e+19                0   
12983         1833     143332      6.841801e+19                0   
12984         1833     143330      6.841801e+19                0   
12985         1833     143330      6.841801e+19                0   
12986         1833     143330      6.841801e+19                0   

      ver_estructura  ...        ide_Ug  persug  Capital  marca  estado  \
0        50118082017  ...  6.841868e+26       3        0      0       0   
1        50118082017  ...  6.841868e+26       5        0      0       0   
2        50118082017  ...  6.841868e+26       5        0      0       0   
3        50118082017  ...  6.841868e+26       5        0      0       0   
4        50118082017  ...  6.841868e+26       5        0      0       0   
...              ...  ...           ...     ...      ...    ...     ...   
12982    50118082017  ...  6.841868e+26       2        0      0       0   
12983    50118082017  ...  6.841868e+26       2        0      0       0   
12984    50118082017  ...  6.841868e+26       3        0      0       0   
12985    50118082017  ...  6.841868e+26       3        0     10      91   
12986    50118082017  ...  6.841868e+26       3        0      0       0   

           Linea_p  replicacion    fec_actualizacion_cns     C  Clasificacion  
0      10781693454            0  2023-11-27 10:33:13.363  4.35            A02  
1      10781693454            0  2023-11-27 10:33:13.363  4.95            C02  
2      10781693454            0  2023-11-27 10:33:13.363  4.95            C02  
3      10781693454            0  2023-11-27 10:33:13.363  4.95            C02  
4      10781693454            0  2023-11-27 10:33:13.363  4.95            C02  
...            ...          ...                      ...   ...            ...  
12982  10781693454            1  2023-12-06 20:11:20.607  0.60            A05  
12983  10781693454            1  2023-12-06 20:11:20.607  0.60            A05  
12984  10781693454            1  2023-12-06 20:11:18.260  3.30            A02  
12985  10781693454            1  2023-12-06 20:11:18.260  3.30            A02  
12986  10781693454            1  2023-12-06 20:11:18.260  3.30            A02  

[12987 rows x 255 columns]

#Columnas a Eliminar#

In [ ]:
def contar(df,feature):
  return df[feature].value_counts()

In [ ]:
for i in range(len(df.columns)):
  print(df.columns[i])
  print(contar(df,df.columns[i]))

cod_dpto
cod_dpto
68    12987
Name: count, dtype: int64
Departamento
Departamento
SANTANDER    12987
Name: count, dtype: int64
cod_mpio
cod_mpio
68418    12987
Name: count, dtype: int64
Municipio
Municipio
LOS SANTOS    12987
Name: count, dtype: int64
fec_paquete
fec_paquete
2023-05-05 17:02:21.163    175
2022-04-07 16:59:58.747    173
2023-04-27 17:15:26.073    110
2023-06-14 17:14:31.410    102
2021-06-04 07:49:31.273    101
                          ... 
2019-09-18 14:04:11.723      1
2022-08-19 13:58:04.660      1
2020-03-06 07:42:16.540      1
2022-09-15 11:04:24.503      1
2023-06-05 17:45:27.880      1
Name: count, Length: 540, dtype: int64
num_paquete
num_paquete
1691    175
1420    173
1681    110
1731    102
1276    101
       ... 
1020      1
1521      1
1121      1
1541      1
1721      1
Name: count, Length: 540, dtype: int64
num_ficha
num_ficha
132645    15
131487    13
143259    13
124744    13
143294    12
          ..
141146     1
140109     1
140111     1
140104     1

*Se eliminan las columnas cuyo valor de conteo sea 12987*


In [ ]:
df_2 = df.drop(["Capital", "ide_Unigasto", "vlr_ingr_remesa_exterior" ,"ide_serv_domestico" , "ind_pariente_domestico", "Ind_declaracion", "cod_critico", "est_ficha", "cod_dpto" ,"Departamento" ,"cod_mpio", "Municipio", "ori_encuesta" ,"Cod_comuna" ,"tot_viviendas" ,"uso_vivienda","Cod_Chip","Dir_Chip", ], axis=1)

##Datos NAN##

In [ ]:
from tabulate import tabulate

In [ ]:
def contar_nan(df,feature):
  return df[feature].isna().sum()

tabla_nan_data=[]
for column in df_2.columns:
  nan_count = contar_nan(df_2,column)
  tabla_nan_data.append([column,nan_count])
tabla_nan_data

headers_nan=["Feature","Cantidad Nan"]
tabla_nan_data=tabulate(tabla_nan_data,headers_nan,tablefmt="fancy_grid")

print(tabla_nan_data)

╒═══════════════════════════╤════════════════╕
│ Feature                   │   Cantidad Nan │
╞═══════════════════════════╪════════════════╡
│ fec_paquete               │              0 │
├───────────────────────────┼────────────────┤
│ num_paquete               │              0 │
├───────────────────────────┼────────────────┤
│ num_ficha                 │              0 │
├───────────────────────────┼────────────────┤
│ ide_ficha_origen          │              0 │
├───────────────────────────┼────────────────┤
│ ide_edificacion           │              0 │
├───────────────────────────┼────────────────┤
│ ver_estructura            │              0 │
├───────────────────────────┼────────────────┤
│ Cod_clase                 │              0 │
├───────────────────────────┼────────────────┤
│ Cod_centro_poblado        │              0 │
├───────────────────────────┼────────────────┤
│ Cod_area_coordinacion     │              0 │
├───────────────────────────┼────────────────┤
│ Cod_area_op

*Limpieza de los datos NAN*

In [ ]:
# Convertir a datetime
df_2['fec_ini_visita2'] = pd.to_datetime(df_2['fec_ini_visita2'], errors='coerce')
df_2['fec_fin_visita2'] = pd.to_datetime(df_2['fec_fin_visita2'], errors='coerce')


## FECHAS DE VISITA ##

#En la visita 1 habian 693 datos tanto en el inc como en el fin

df_2['visita1_realizada'] = df_2['fec_ini_visita1'].notna().astype(int)

#En la visita 2 habian mas datos NAN en la fecha de inc
df_2['visita2_realizada'] = df_2[['fec_ini_visita2', 'fec_fin_visita2']].notna().any(axis=1).astype(int)
df_2['fec_ini_visita2'] = df_2['fec_ini_visita2'].fillna(df_2['fec_fin_visita2'])



In [ ]:
## DATOS PERSONALES ##

df_2['seg_nom_informante'] = df_2['seg_nom_informante'].fillna('NO TIENE')
df_2['seg_ape_informante'] = df_2['seg_ape_informante'].fillna('NO TIENE')

df_2['seg_apellido'] = df_2['seg_apellido'].fillna('NO TIENE')
df_2['seg_nombre'] = df_2['seg_nombre'].fillna('NO TIENE')


In [ ]:
## FIRMA DEL INFROMANTE ##

df_2['informante_firmo'] = df_2['ide_firma_informante'].notna().astype(int)
df_2['ide_firma_informante'] = df_2['ide_firma_informante'].fillna('NO FIRMÓ')


In [ ]:
## EMAIL ##

df_2['tiene_email'] = df_2['Email_contacto'].notna().astype(int)
df_2['Email_contacto'] = df_2['Email_contacto'].fillna('NO TIENE CORREO')


In [ ]:
## JEFE UG##

df_2['jefe_ug_identificado'] = df_2['Jefe_UG'].apply(lambda x: 1 if x == 1 else 0)

In [ ]:
## ACTUALIZACION CNS ##
df_2['cns_actualizado'] = df_2['fec_actualizacion_cns'].notna().astype(int)
df_2['fec_actualizacion_cns'] = df_2['fec_actualizacion_cns'].fillna('NO ACTUALIZADO')

In [ ]:
## PERSONA 1  ##
df_2['persona1_encuestada'] = df_2['fec_ini_persona1'].notna().astype(int)

In [ ]:
df_2[['visita1_realizada', 'visita2_realizada', 'informante_firmo', 'tiene_email', 'cns_actualizado']].head(15)

visita1_realizada  visita2_realizada  informante_firmo  tiene_email  \
0                   1                  1                 1            0   
1                   1                  1                 1            0   
2                   1                  1                 1            0   
3                   1                  1                 1            0   
4                   1                  1                 1            0   
5                   1                  1                 1            0   
6                   1                  1                 1            0   
7                   1                  1                 1            0   
8                   1                  0                 1            0   
9                   1                  0                 1            0   
10                  1                  0                 1            0   
11                  1                  0                 1            0   
12                  1                  0                 1            0   
13                  1                  1                 1            0   
14                  1                  0                 1            0   

    cns_actualizado  
0                 1  
1                 1  
2                 1  
3                 1  
4                 1  
5                 1  
6                 1  
7                 1  
8                 1  
9                 1  
10                1  
11                1  
12                1  
13                1  
14                1

In [ ]:
df_2.columns[df_2.isna().any()]


Index(['fec_ini_visita1', 'fec_fin_visita1', 'fec_ini_visita2',
       'fec_fin_visita2', 'fec_ini_persona1', 'fec_fin_persona1', 'Jefe_UG'],
      dtype='object')

#Parte 2 Analisis de Clusters#


In [ ]:
df_2.columns

Index(['fec_paquete', 'num_paquete', 'num_ficha', 'ide_ficha_origen',
       'ide_edificacion', 'ver_estructura', 'Cod_clase', 'Cod_centro_poblado',
       'Cod_area_coordinacion', 'Cod_area_operativa',
       ...
       'fec_actualizacion_cns', 'C', 'Clasificacion', 'visita1_realizada',
       'visita2_realizada', 'informante_firmo', 'tiene_email',
       'jefe_ug_identificado', 'cns_actualizado', 'persona1_encuestada'],
      dtype='object', length=244)

In [ ]:
variables_funcionales = [
    'ind_ingr_cosecha', 'ind_ingr_salario', 'ind_ingr_estado',
    'ind_ingr_honorarios', 'ind_ingr_pension', 'ind_fondo_pensiones',
    'C', 'edad_calculada', 'tip_empleado'
]


df_cluster = df_2[variables_funcionales].copy()
df_cluster = pd.get_dummies(df_cluster, columns=['tip_empleado'], prefix='empleo')

df_cluster.columns.tolist()

['ind_ingr_cosecha',
 'ind_ingr_salario',
 'ind_ingr_estado',
 'ind_ingr_honorarios',
 'ind_ingr_pension',
 'ind_fondo_pensiones',
 'C',
 'edad_calculada',
 'empleo_1',
 'empleo_2',
 'empleo_3',
 'empleo_4',
 'empleo_5',
 'empleo_6',
 'empleo_7',
 'empleo_8',
 'empleo_9',
 'empleo_10',
 'empleo_99']

| Código | Ocupación                                                        | Zona asignada            | Justificación                                                                |
| ------ | ---------------------------------------------------------------- | ------------------------ | ---------------------------------------------------------------------------- |
| 1      | Empleado de empresa particular                                   | Urbana                   | Trabajo asalariado no rural, alta probabilidad de estar en ciudad.           |
| 2      | Empleado del gobierno                                            | Urbana                   | Claramente urbano, en sector público.                                        |
| 3      | Empleado doméstico                                               | Urbana                   | Actividad asociada a hogares urbanos.                                        |
| 4      | Profesional independiente                                        | Urbana                   | Generalmente ubicado en zona urbana.                                         |
| 5      | Trabajador independiente o por cuenta propia                     | Urbana                   | Puede ser mixto, pero lo inferimos como urbano por descarte.                 |
| 6      | Patrón o empleador                                               | **Avícola**              | Lo inferimos como productor avícola en ausencia de ingresos agrícolas.       |
| 7      | Trabajador de finca, tierra o parcela propia                     | Agrícola                 | Trabajo directamente ligado a actividades agrícolas.                         |
| 8      | Trabajador sin remuneración                                      | **Avícola**              | Lo inferimos como ayudante o trabajador familiar en finca avícola.           |
| 9      | Ayudante sin remuneración (hijo o familiar de empleados rurales) | **Avícola**              | Similar al 8, se infiere que apoya labores productivas rurales no agrícolas. |
| 10     | Jornalero o peón                                                 | Agrícola                 | Típicamente vinculado a labores agrícolas.                                   |
| 99     | No aplica                                                        | NaN (o "Sin clasificar") | No hay forma de inferir.                                                     |


In [ ]:
UMBRAL_URBANO = df['vlr_ingr_salario'].quantile(0.75)


In [ ]:
UMBRALMEDIO_URBANO = df['vlr_ingr_salario'].quantile(0.5)

In [ ]:
def asignar_zona(row):
    salario = row['vlr_ingr_salario']
    ingr_agricola = row['vlr_ingr_cosecha']
    emp = row['tip_empleado']

    # 1. Agrícola
    if ingr_agricola > 0 or emp in [7, 10]:
        return 'Agrícola'

    # 2. Urbana
    if emp in [2, 4, 5]:  # Ajusta el umbral a lo que consideres alto
        return 'Urbana'

    # 3. Mineral
    if emp == 1 and salario > UMBRAL_URBANO:
        return 'Mineral'

    # 4. Avícola
    if emp in [6, 8, 9]:
        return 'Avícola'

    # 5. No clasificados
    return np.nan

# Aplicar al DataFrame
df['zona_final'] = df.apply(asignar_zona, axis=1)


In [ ]:
print(df['zona_final'].value_counts())


zona_final
Agrícola    1350
Urbana       582
Mineral      401
Avícola       31
Name: count, dtype: int64


In [ ]:
import pandas as pd
from tabulate import tabulate  # opcional para imprimir bonito en consola

# Función para construir el perfil por zona con tus variables corregidas
def construir_perfil(df):
    perfil = df.groupby('zona_final').agg(
        Total=('zona_final', 'count'),
        Edad_Promedio=('edad_calculada', 'mean'),
        Porc_Mujeres=('sexo_persona', lambda x: (x == 2).mean() * 100),
        Porc_Educ_Baja=('niv_educativo', lambda x: (x.isin([0, 1, 2])).mean() * 100),
        Porc_Pobreza_Multidimensional=('C', lambda x: (x > 5).mean() * 100),
        Salario_Promedio=('vlr_ingr_salario', 'mean')
    ).round(2)

    # Formatear columnas para presentación
    perfil['Edad_Promedio'] = perfil['Edad_Promedio'].astype(int)
    perfil['Salario_Promedio'] = perfil['Salario_Promedio'].apply(lambda x: f"${x:,.0f}")
    perfil['Porc_Mujeres'] = perfil['Porc_Mujeres'].astype(str) + "%"
    perfil['Porc_Educ_Baja'] = perfil['Porc_Educ_Baja'].astype(str) + "%"
    perfil['Porc_Pobreza_Multidimensional'] = perfil['Porc_Pobreza_Multidimensional'].astype(str) + "%"

    return perfil

# Ejecutar función
tabla_perfil = construir_perfil(df)

# Mostrar bonito en Jupyter
display(tabla_perfil)

# Mostrar bonito en consola
print(tabulate(tabla_perfil, headers='keys', tablefmt='fancy_grid'))


Total  Edad_Promedio Porc_Mujeres Porc_Educ_Baja  \
zona_final                                                     
Agrícola     1350             41       14.89%          64.0%   
Avícola        31             36       54.84%         58.06%   
Mineral       401             39        18.7%         46.38%   
Urbana        582             41       34.36%         44.85%   

           Porc_Pobreza_Multidimensional Salario_Promedio  
zona_final                                                 
Agrícola                          23.85%         $192,702  
Avícola                           35.48%               $0  
Mineral                            5.49%         $596,313  
Urbana                            15.64%          $74,739

╒══════════════╤═════════╤═════════════════╤════════════════╤══════════════════╤═════════════════════════════════╤════════════════════╕
│ zona_final   │   Total │   Edad_Promedio │ Porc_Mujeres   │ Porc_Educ_Baja   │ Porc_Pobreza_Multidimensional   │ Salario_Promedio   │
╞══════════════╪═════════╪═════════════════╪════════════════╪══════════════════╪═════════════════════════════════╪════════════════════╡
│ Agrícola     │    1350 │              41 │ 14.89%         │ 64.0%            │ 23.85%                          │ $192,702           │
├──────────────┼─────────┼─────────────────┼────────────────┼──────────────────┼─────────────────────────────────┼────────────────────┤
│ Avícola      │      31 │              36 │ 54.84%         │ 58.06%           │ 35.48%                          │ $0                 │
├──────────────┼─────────┼─────────────────┼────────────────┼──────────────────┼─────────────────────────────────┼────────────────────┤
│ Mineral      │     401 │              39 │ 18.